utoipa/src/openapi/extensions.rs

```diff
@@ -0,0 +1,97 @@
+//! Implements [OpenAPI Extensions][external_docs].
+//!
+/// [extensions]: https://spec.openapis.org/oas/latest.html#specification-extensions
+use std::{
+    collections::HashMap,
+    ops::{Deref, DerefMut},
+};
+
+use serde::Serialize;
+
+use super::builder;
+
+const EXTENSION_PREFIX: &str = "x-";
+
+builder! {
+    ExtensionsBuilder;
+
+    /// Additional [data for extending][extensions] the OpenAPI specification.
+    ///
+    /// [extensions]: https://spec.openapis.org/oas/latest.html#specification-extensions
+    #[derive(Default, Serialize, Clone, PartialEq, Eq)]
+    #[cfg_attr(feature = "debug", derive(Debug))]
+    pub struct Extensions{
+        #[serde(flatten)]
+        extensions: HashMap<String, serde_json::Value>,
+    }
+}
+
+impl Deref for Extensions {
+    type Target = HashMap<String, serde_json::Value>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.extensions
+    }
+}
+
+impl DerefMut for Extensions {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.extensions
+    }
+}
+
+impl From<Extensions> for HashMap<String, serde_json::Value> {
+    fn from(value: Extensions) -> Self {
+        value.extensions
+    }
+}
+
+impl<'de> serde::de::Deserialize<'de> for Extensions {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let extensions: HashMap<String, _> = HashMap::deserialize(deserializer)?;
+        let extensions = extensions
+            .into_iter()
+            .filter(|(k, _)| k.starts_with(EXTENSION_PREFIX))
+            .collect();
+        Ok(Self { extensions })
+    }
+}
+
+impl ExtensionsBuilder {
+    /// Adds a key-value pair to the extensions. Extensions keys are prefixed with `"x-"` if
+    /// not done already.
+    pub fn add<K, V>(mut self, key: K, value: V) -> Self
```

Comment:
I'm unsure about the `Extensions` public API. Should we expose the underlying `HashMap`? Or should it only be accessed via setters/getters to ensure that the extensions keys have the correct format?

Reply:
I think that is good for the builder to have this kind of adding method. However as what comes to the `Extension` type itself I think it is good to follow the common practice overall and just expose the fields as public like other types does. 

And for `Extension` there is nothing to expose as it is exposed via `Deref` and `DerefMut` directly so I think this is good as is. When the field is exposed as is it allows users to move vast approaches to operate on the underlying type. Though with the expense of not enforcing the `x- prefix for keys` as defined by the specification.

Reply:
As a second though I have no strict opinion whether the `getter/setter` should be enforced for the `Extension` type as well. At least then it could enforce the `x-` prefix, but at the same time is limiting how users can actually operate on the type. E.g. I believe it would be useful for example to allow adding extensions from `iter` as well directly to the map and to be able to operate directly with the map via Rust's Iterator API.